### DISCUSSION TITLE
Not awaiting async function inside of non-edge API route/server actions.

### DISCUSSION THREAD

**Discussion by max-widmaier** [vercel/next.js#66306] [Help]
### Summary

So, I've read a bit about how you have to await everything in an edge function because they run on Lambda which die shortly after the function returns. Does this also apply to server actions and API routes?
I guess this is really a question of infrastructure when deploying on Vercel. Locally, I can do this just fine. 
When I deploy to Vercel, will the Next server run distributed or as an actual server/cluster? Can I specify that the various server actions and API routes _not_ run on Edge.
Even if Next deploys to a server cluster, will it kill processes even if there are unresolved promises?

I've provided some code snippets which are more or less pseudo-code for use cases in the codebase I'm working on.

TL;DR: Do you have to await all async functions in server actions and API routes that do not run on the Edge?

### Additional information

```js
// Server action
export async function uploadFile(file: File, fileName: string) {
    if (await isAuthenticated()) {
        const client = new S3Client({region: process.env.AWS_REGION});
        const command = new PutObjectCommand({
            Bucket: "test-bucket",
            Key: fileName,
            Body: file,
         });
        client.send(command); // Not awaited

        log(`someone upload a file name ${fileName}`); // Also not awaited

        return `Woohoo, you uploaded ${fileName}`;
    }

    return "Log in please";
}

// Update settings API route
export async function POST(req: NextRequest) {
    let user: User = await getUser();
    if (user.isAdmin()) {
        let {key, value} = await req.body();
        await prisma.settings.update({
            where: { key },
            data: { value }
        });

        log(`${user.name} updated ${key} to ${value}`); // Not awaited
    }
}
```

**Comment by gregauger** [comment]
The answer here is `after`: https://nextjs.org/docs/app/api-reference/functions/after

**Comment by lizuju** [comment]
I would split this into two cases: work that must complete before you can honestly return success, and side effects that are allowed to happen after the response.

For the S3 example, the upload is not a background side effect. It is the main mutation. I would await it:

```ts
await client.send(command)
return `Woohoo, you uploaded ${fileName}`
```

If you do not await it, the action can return success while the upload is still pending or has already failed. On Vercel, even in the Node.js runtime, you should still treat the invocation as request-scoped. It is not a durable always-on worker where floating promises are a reliable contract.

For logging/analytics/non-critical cleanup, use `after()`:

```ts
import { after } from 'next/server'

export async function POST(req: NextRequest) {
  const user = await getUser()

  if (!user.isAdmin()) {
    return Response.json({ error: 'Unauthorized' }, { status: 401 })
  }

  const { key, value } = await req.json()

  await prisma.settings.update({
    where: { key },
    data: { value },
  })

  after(() => {
    log(`${user.name} updated ${key} to ${value}`)
  })

  return Response.json({ ok: true })
}
```

`after()` is the right primitive because Next/Vercel can keep the request context alive for that scheduled work after the response is finished. But I would not use it for critical work that needs retries, durability, or long execution time. For that, push a job to a queue and return after the job is accepted.

So the short version is:

- DB writes, uploads, payments, emails you must guarantee: `await` them, or enqueue them durably and await the enqueue.
- Logs, metrics, analytics, cache cleanup: `after()` is fine.
- Plain unawaited promises: avoid relying on them in Route Handlers or Server Actions, even outside Edge.

**Comment by Hassanmahmood4** [comment]
**Short answer:** Yes, you should await everything — even outside of Edge. Not awaiting is a bug waiting to happen on Vercel (and most production deployments).

## Why it matters

It comes down to how Vercel runs your Next.js app. Even when you're not using Edge Runtime, your server actions and API routes run as **serverless functions** (Node.js Lambda) by default on Vercel. The lifecycle is:

1. Request comes in → cold start or reuse a warm function
2. Your handler runs → returns a response
3. **Function freezes** → the process isn't killed, but it's frozen in place. CPU is paused. Event loop stops.

That third step is the problem. When you fire off `client.send(command)` without awaiting it, your function returns the response, Vercel sees "this function is done," and freezes the process. Your S3 upload is mid-flight — the TCP connection is open, bytes are partially sent — and everything just stops. It *might* complete on the next invocation when the function unfreezes, or it might not. It's completely non-deterministic.

## What happens to your code specifically

```js
client.send(command); // Not awaited
log(`someone uploaded a file named ${fileName}`); // Not awaited
return `Woohoo, you uploaded ${fileName}`;
```

The user gets "Woohoo" back immediately. But:
- The S3 upload may or may not complete — depends on whether the Lambda stays warm and unfreezes before the connection times out
- The log call has the same problem
- Locally it works fine because your Node process is long-lived and the event loop keeps running after the response is sent

## The fix

Await everything that has a side effect:

```js
export async function uploadFile(file: File, fileName: string) {
    if (await isAuthenticated()) {
        const client = new S3Client({ region: process.env.AWS_REGION });
        const command = new PutObjectCommand({
            Bucket: "test-bucket",
            Key: fileName,
            Body: file,
        });
        await client.send(command);
        // don't await log if you truly don't care whether it succeeds,
        // but do run it concurrently so it finishes before the function exits:
        await log(`someone uploaded a file named ${fileName}`);
        return `Woohoo, you uploaded ${fileName}`;
    }
    return "Log in please";
}
```

If you want the S3 upload and the log to run concurrently (faster), use `Promise.all`:

```js
await Promise.all([
    client.send(command),
    log(`someone uploaded a file named ${fileName}`),
]);
```

This still awaits both — you just don't wait for them sequentially.

## For your API route — same thing

```js
export async function POST(req: NextRequest) {
    let user: User = await getUser();
    if (user.isAdmin()) {
        let { key, value } = await req.body();
        await prisma.settings.update({
            where: { key },
            data: { value },
        });
        await log(`${user.name} updated ${key} to ${value}`);
    }
}
```

## Can you opt out of serverless?

You can force specific routes to use Node.js runtime (not Edge) with:

```js
export const runtime = 'nodejs'; // default, not edge
```

But on Vercel, **even `nodejs` runtime routes are serverless functions** — they still freeze after the response. The only way to get a persistent long-lived process on Vercel is to use a custom server setup, which defeats the purpose of using Vercel.

If you truly need fire-and-forget background work (like sending emails, processing images after responding to the user), the correct patterns are:
- **Vercel Functions with `waitUntil`** — extends the function lifetime after the response:
```js
  import { waitUntil } from '@vercel/functions';

  export async function POST(req: NextRequest) {
      // respond immediately
      waitUntil(log(`something happened`)); // runs after response, but Vercel waits for it
      return NextResponse.json({ ok: true });
  }
```
- **A queue** (SQS, Inngest, Trigger.dev) — push the work to a queue, return immediately, process it separately

## TL;DR

| Environment | Un-awaited promises safe? |
|---|---|
| Local dev (`next dev`) | Works (process stays alive) |
| Vercel — Edge | No (Lambda dies) |
| Vercel — Node.js | No (Lambda freezes) |
| Self-hosted (`next start` on a VPS) | Works (persistent process) |
| Docker / long-running server | Works (persistent process) |

Always await. If you need to respond before async work finishes, use `waitUntil` or a queue.

**Reply by Hassanmahmood4** [reply]
Could you mark the answer as accepted so others can find it easily?"

**Comment by webdevpraveen** [comment]
To answer your question directly: **Yes, this will absolutely cause random failures in production on Vercel (Serverless).**

Here is why it works for you right now, but will fail later:
When you run `next dev` locally, or if you deploy to a long-running Node.js server (like an EC2 instance or Docker container), the Node process stays alive indefinitely. Background async functions will happily finish executing even after the HTTP response is sent.

However, when deployed as Serverless Functions (which is what non-edge API routes and Server Actions compile down to on Vercel), the execution context is "frozen" or completely destroyed the moment the main request handler resolves and returns the response. Any unawaited promises floating in the background will be killed midway.

### The Modern Solution
If you want to return a fast response to the user but keep a background task running (like sending an email or pinging a 3rd party API), you should not leave it unawaited. Instead, use the `waitUntil()` API provided by Vercel.

```ts
import { waitUntil } from '@vercel/functions';

export async function myServerAction() {
  // Pass the unawaited promise to waitUntil
  waitUntil(
    fetch('https://3rd-party-api.com/log', { method: 'POST' })
  );
  
  // Return immediately to the user
  return { success: true };
}
